src/main/kotlin/org/rust/lang/core/resolve/KnownItems.kt

```diff
@@ -0,0 +1,127 @@
+/*
+ * Use of this source code is governed by the MIT license that can be
+ * found in the LICENSE file.
+ */
+
+package org.rust.lang.core.resolve
+
+import com.intellij.openapi.project.Project
+import com.intellij.openapi.util.Key
+import com.intellij.reference.SoftReference
+import com.intellij.util.containers.ContainerUtil
+import org.rust.cargo.project.model.CargoProject
+import org.rust.cargo.project.workspace.CargoWorkspace
+import org.rust.cargo.util.AutoInjectedCrates.CORE
+import org.rust.lang.core.psi.RsTraitItem
+import org.rust.lang.core.psi.ext.RsElement
+import org.rust.lang.core.psi.ext.RsNamedElement
+import org.rust.lang.core.psi.ext.RsStructOrEnumItemElement
+import org.rust.lang.core.psi.ext.cargoProject
+import org.rust.lang.core.resolve.indexes.RsLangItemIndex
+import org.rust.openapiext.getOrPutSoft
+import java.util.*
+
+private val KNOWN_ITEMS_KEY: Key<SoftReference<KnownItems>> = Key.create("KNOWN_ITEMS_KEY")
+private val DUMMY_KNOWN_ITEMS = KnownItems(DummyKnownItemsLookup)
+
+val CargoProject.knownItems: KnownItems
+    get() = getOrPutSoft(KNOWN_ITEMS_KEY) {
+        val workspace = workspace ?: return@getOrPutSoft DUMMY_KNOWN_ITEMS
+        KnownItems(RealKnownItemsLookup(intellijProject, workspace))
+    }
+
+val RsElement.knownItems: KnownItems
+    get() = cargoProject?.knownItems ?: DUMMY_KNOWN_ITEMS
+
+/**
+ * [CargoProject]-related object that allows to lookup rust items in project dependencies,
+ * including the standard library.
+ */
+@Suppress("PropertyName")
+class KnownItems(
```

Comment:
I'm still not sure that it's the best name but I can't come up with a better one so I think it's ok for now